app/tests/spec/lib/validate.js

```diff
@@ -134,5 +134,112 @@ define(function (require, exports, module) {
         assert.isTrue(Validate.isPasswordValid(createRandomHexString(Constants.PASSWORD_MIN_LENGTH)));
       });
     });
+
+    describe('isDataValid', function () {
+      /*eslint-disable sorting/sort-object-props */
+
+      it('is defined', function () {
+        assert.isFunction(Validate.isDataValid);
+      });
+
+      it('expects two arguments', function () {
+        assert.lengthOf(Validate.isDataValid, 2);
+      });
+
+      it('returns true when simple type matches the schema', function () {
+        assert.isTrue(Validate.isDataValid('foo', 'String'));
+      });
+
+      it('returns false when simple type does not match the schema', function () {
+        assert.isFalse(Validate.isDataValid('foo', 'Number'));
+      });
+
+      it('handles null', function () {
+        assert.isTrue(Validate.isDataValid(null, 'Null'));
+      });
+
+      it('handles undefined', function () {
+        assert.isTrue(Validate.isDataValid(undefined, 'Undefined'));
+      });
+
+      it('handles false', function () {
+        assert.isTrue(Validate.isDataValid(false, 'Boolean'));
+      });
+
+      it('handles zero', function () {
+        assert.isTrue(Validate.isDataValid(0, 'Number'));
+      });
+
+      it('handles empty string', function () {
+        assert.isTrue(Validate.isDataValid('', 'String'));
+      });
+
+      it('returns true when complex type matches the schema', function () {
+        assert.isTrue(Validate.isDataValid(
+          [ 'foo', 3.14159265359, {} ],
+          [ 'String', 'Number', 'Object' ]
+        ));
+      });
+
+      it('returns false when complex type does not match the schema', function () {
+        assert.isFalse(Validate.isDataValid(
+          [ 'foo', 3.14159265359, [] ],
+          [ 'String', 'Number', 'Object' ]
```

Comment:
> Can you make a proposal how we'd handle [sending sync data](https://github.com/mozilla/fxa-content-server/blob/2d7c9e2afe7ad9df0f721ba527615f5b420521ea/app/scripts/models/auth_brokers/fx-sync.js#L207-L221)...

I can think of a few:

1. Validate `declinedSyncEngines` as `'Array'` and be content with rudimentary validation for that field. Mentioning this one explicitly as I don't know what the risks/rewards are around validating more strictly (or not) here.

   ```js
   if (isDataValid(loginData, { ..., declinedSyncEngines: 'Array' }) {
   ```

2. Add a separate validation step for that field then implement `isArrayValid` to iterate through the array and apply the supplied schema on each value.

   ```js
   if (isDataValid(loginData, { ..., declinedSyncEngines: 'Array' }) &&
     isArrayValid(loginData.declinedSyncEngines, 'String')) {
   ```

3. Encode the item type in the schema syntax then separate array and object processing inside `isDataValid`. This is a more limited solution than `2` because it doesn't permit the validation of complex array items.

   ```js
   if (isDataValid(loginData, { ..., declinedSyncEngines: 'Array-String' }) {
   ```

4. A variation on `3`:

   ```js
   if (isDataValid(loginData, { ..., declinedSyncEngines: [ 'String' ] }) {
   ```

Option `1` or `2` seems best to me. Have you got any other suggestions that I've not considered?